runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -1171,6 +1173,8 @@ private void cancelPendingTimerById(String contextTimerId) throws Exception {
     void cleanupPendingTimer(TimerData timer) {
       try {
         pendingTimersById.remove(getContextTimerId(timer.getTimerId(), timer.getNamespace()));
+        // Always update watermark hold after a timer has been fired or removed
+        updateWatermarkHold();
```

Comment:
This was the missing piece. Without this, we can't make progress when a timer has fired.

Reply:
Thanks!  I didn't put a call here, because the way the code is structured that would result in two calls to updateWatermarkHold every time you set a timer. Would it be sufficient to add the call in fireTimer, as it appears that's that path we missed.

Reply:
Yes, that's right. My first thought was to add it in `fireTimer`, but then I thought I could simplify it further. That led to the duplicate update call when you set the timer. Looks like we have just update the hold on `set`, `delete`, and `fire`, and not for `cleanup`.